#Cleaning a messy company employee dataset

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv(r"/content/drive/MyDrive/Colab Notebooks/IDRA/Day11_Messy_Company_Employee_Dataset.csv")

In [3]:
df = pd.DataFrame(df)
df.head()

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,NaN,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,NaN,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Employee_ID        157 non-null    object 
 1   Employee_Name      157 non-null    object 
 2   Department         152 non-null    object 
 3   Job_Title          157 non-null    object 
 4   Age                153 non-null    float64
 5   Gender             152 non-null    object 
 6   Annual_Salary      152 non-null    float64
 7   Experience_Years   154 non-null    float64
 8   Joining_Date       157 non-null    object 
 9   City               152 non-null    object 
 10  Performance_Score  154 non-null    float64
 11  Work_Mode          155 non-null    object 
dtypes: float64(4), object(8)
memory usage: 14.8+ KB


There are 157 rows. It can be seen that there are
- 5 null values in the column 'Department'
- 4 null values in the column 'Age'
- 5 null values in the column 'Gender'
- 5 null values in the column 'Annual_Salary'
- 3 null values in the column 'Experience_Years'
- 5 null values in the column 'City'
- 3 null values in the column 'Performance_Score'
- 2 null values in the column 'Work_Mode'

This can be automatically generated using *isnull().sum()*

In [5]:
df.isnull().sum()

,0
Employee_ID,0
Employee_Name,0
Department,5
Job_Title,0
Age,4
Gender,5
Annual_Salary,5
Experience_Years,3
Joining_Date,0
City,5


Total no. of null values can be seen using *isnull().sum().sum()*

In [6]:
df.isnull().sum().sum()

np.int64(32)

#Dealing with the null values

In [7]:
df['Department'] = df['Department'].fillna(df['Department'].mode()[0])
df['Age'] = df['Age'].fillna(df['Age'].median())
df['Gender'] = df['Gender'].fillna(df['Gender'].mode()[0])
df['Annual_Salary'] = df['Annual_Salary'].fillna(df['Annual_Salary'].median())
df['Experience_Years'] = df['Experience_Years'].fillna(df['Experience_Years'].median())
df['Performance_Score'] = df['Performance_Score'].fillna(df['Performance_Score'].median())
df['City'] = df['City'].fillna(df['City'].mode()[0])
df['Work_Mode'] = df['Work_Mode'].fillna(df['Work_Mode'].mode()[0])

#Inconsistent entries

In [8]:
print(df['Department'].unique())
print(df['Gender'].unique())
print(df['City'].unique())
print(df['Work_Mode'].unique())

['Marketing' 'Human Resources' 'Engineering' 'Sales' 'Finance'
 'Data & Analytics' 'Customer Success' 'Operations' 'ENGINEERING'
 'engineering']
['Other' 'Female' 'Male' 'female' 'MALE']
['Pune' 'Mumbai' 'Bengaluru' 'Jaipur' 'Hyderabad' 'Chennai' 'Chandigarh'
 'Srinagar' 'Kolkata' 'delhi' 'Delhi' 'DELHI' 'Delhi ']
['Office' 'Remote' 'Hybrid' 'REMOTE' 'remote']


In [9]:
df['City'] = df['City'].str.strip()
print(df['City'].unique())

['Pune' 'Mumbai' 'Bengaluru' 'Jaipur' 'Hyderabad' 'Chennai' 'Chandigarh'
 'Srinagar' 'Kolkata' 'delhi' 'Delhi' 'DELHI']


In [10]:
df['Department'] = df['Department'].str.title()
df['Gender'] = df['Gender'].str.title()
df['City'] = df['City'].str.title()
df['Work_Mode'] = df['Work_Mode'].str.title()

print(df['Department'].unique())
print(df['Gender'].unique())
print(df['City'].unique())
print(df['Work_Mode'].unique())

['Marketing' 'Human Resources' 'Engineering' 'Sales' 'Finance'
 'Data & Analytics' 'Customer Success' 'Operations']
['Other' 'Female' 'Male']
['Pune' 'Mumbai' 'Bengaluru' 'Jaipur' 'Hyderabad' 'Chennai' 'Chandigarh'
 'Srinagar' 'Kolkata' 'Delhi']
['Office' 'Remote' 'Hybrid']


In [11]:
print(df.dtypes)

Employee_ID           object
Employee_Name         object
Department            object
Job_Title             object
Age                  float64
Gender                object
Annual_Salary        float64
Experience_Years     float64
Joining_Date          object
City                  object
Performance_Score    float64
Work_Mode             object
dtype: object


In [12]:
df['Age'] = pd.to_numeric(df['Age'])
df['Experience_Years'] = pd.to_numeric(df['Experience_Years'])
df['Annual_Salary'] = pd.to_numeric(df['Annual_Salary'])
df['Performance_Score'] = pd.to_numeric(df['Performance_Score'])
df['Joining_Date'] = pd.to_datetime(df['Joining_Date'])

In [13]:
df.head()

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,Marketing,Data Scientist,48.0,Other,108371.0,13.20,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.10,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.20,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,9.35,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.90,2018-07-24,Hyderabad,4.0,Hybrid


In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Employee_ID        157 non-null    object        
 1   Employee_Name      157 non-null    object        
 2   Department         157 non-null    object        
 3   Job_Title          157 non-null    object        
 4   Age                157 non-null    float64       
 5   Gender             157 non-null    object        
 6   Annual_Salary      157 non-null    float64       
 7   Experience_Years   157 non-null    float64       
 8   Joining_Date       157 non-null    datetime64[ns]
 9   City               157 non-null    object        
 10  Performance_Score  157 non-null    float64       
 11  Work_Mode          157 non-null    object        
dtypes: datetime64[ns](1), float64(4), object(7)
memory usage: 14.8+ KB


#Finding duplicates

In [15]:
df.duplicated().sum()

np.int64(7)

In [16]:
df[df.duplicated(keep=False)]

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
11,EMP0075,Manya Joshi,Sales,Sales Manager,29.0,Male,92623.0,11.9,2025-08-17,Jaipur,4.0,Hybrid
25,EMP0097,Zoya Shah,Finance,Financial Analyst,41.0,Male,81531.0,4.1,2023-06-05,Mumbai,4.0,Office
55,EMP0097,Zoya Shah,Finance,Financial Analyst,41.0,Male,81531.0,4.1,2023-06-05,Mumbai,4.0,Office
61,EMP0075,Manya Joshi,Sales,Sales Manager,29.0,Male,92623.0,11.9,2025-08-17,Jaipur,4.0,Hybrid
70,EMP0021,Riya Khan,Operations,Operations Executive,49.0,Female,71549.0,4.6,2024-09-20,Chandigarh,3.0,Hybrid
78,EMP0048,Nikhil Dar,Finance,Financial Analyst,27.0,Female,90046.5,10.1,2020-06-18,Chandigarh,2.0,Office
85,EMP0139,Simran Sharma,Finance,Accountant,32.0,Female,110617.0,10.3,2021-08-03,Mumbai,3.0,Office
92,EMP0120,Saira Sheikh,Marketing,Marketing Executive,52.0,Male,107328.0,17.4,2020-11-25,Bengaluru,2.0,Remote
100,EMP0139,Simran Sharma,Finance,Accountant,32.0,Female,110617.0,10.3,2021-08-03,Mumbai,3.0,Office
104,EMP0021,Riya Khan,Operations,Operations Executive,49.0,Female,71549.0,4.6,2024-09-20,Chandigarh,3.0,Hybrid


In [17]:
#dropping the duplicated rows
df.drop_duplicates(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150 entries, 0 to 155
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Employee_ID        150 non-null    object        
 1   Employee_Name      150 non-null    object        
 2   Department         150 non-null    object        
 3   Job_Title          150 non-null    object        
 4   Age                150 non-null    float64       
 5   Gender             150 non-null    object        
 6   Annual_Salary      150 non-null    float64       
 7   Experience_Years   150 non-null    float64       
 8   Joining_Date       150 non-null    datetime64[ns]
 9   City               150 non-null    object        
 10  Performance_Score  150 non-null    float64       
 11  Work_Mode          150 non-null    object        
dtypes: datetime64[ns](1), float64(4), object(7)
memory usage: 15.2+ KB


we can see that the 7 duplicate rows are dropped/deleted

#Re-checking

In [18]:
df.isnull().sum().sum()

np.int64(0)

In [19]:
print(df['Department'].unique())
print(df['Gender'].unique())
print(df['City'].unique())
print(df['Work_Mode'].unique())

['Marketing' 'Human Resources' 'Engineering' 'Sales' 'Finance'
 'Data & Analytics' 'Customer Success' 'Operations']
['Other' 'Female' 'Male']
['Pune' 'Mumbai' 'Bengaluru' 'Jaipur' 'Hyderabad' 'Chennai' 'Chandigarh'
 'Srinagar' 'Kolkata' 'Delhi']
['Office' 'Remote' 'Hybrid']


In [20]:
df.duplicated().sum()

np.int64(0)

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150 entries, 0 to 155
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Employee_ID        150 non-null    object        
 1   Employee_Name      150 non-null    object        
 2   Department         150 non-null    object        
 3   Job_Title          150 non-null    object        
 4   Age                150 non-null    float64       
 5   Gender             150 non-null    object        
 6   Annual_Salary      150 non-null    float64       
 7   Experience_Years   150 non-null    float64       
 8   Joining_Date       150 non-null    datetime64[ns]
 9   City               150 non-null    object        
 10  Performance_Score  150 non-null    float64       
 11  Work_Mode          150 non-null    object        
dtypes: datetime64[ns](1), float64(4), object(7)
memory usage: 15.2+ KB


The dataset is now cleaned

**Before**
- null values present in various columns
- repeated entries (same word in different writing styles)
- duplicate rows present
- incorrect column datatypes

**After**
- null values are replaced with some value using median or mode, according to the column type
- repeated categorical words are cleaned
- duplicate rows are removed
- column datatypes are corrected

#Exporting the cleaned dataset

In [22]:
df.to_csv(r'/content/drive/MyDrive/Colab Notebooks/IDRA/Day11_Cleaned_Company_Employee_Dataset.csv', index=False)